include/seqan3/core/algorithm/configurator.hpp

```diff
@@ -0,0 +1,596 @@
+
+namespace seqan3::detail
+{
+// ----------------------------------------------------------------------------
+// configurator_base
+// ----------------------------------------------------------------------------
+
+/*!\brief Abstract base class for the seqan3::detail::configurator class.
+ * \tparam configs_t Template parameter pack over all configs. These types must satisfy the
+ *                   seqan3::detail::config_concept.
+ * \ingroup core_algorithm
+ *
+ * \details
+ *
+ * This is an abstract base class adding a constructor to the seqan3::detail::configurator class with special copy semantics:
+ * When a configurator is extended by a new config type it must be placed at the beginning. To copy the states of
+ * the old configurator, i.e. it contains the states of the already contained config types, the constructors of the
+ * old config types must be called with the previous configurator to properly submit the states to the extended
+ * configurator instance.
+ */
+
+template <typename ... configs_t>
+class configurator_base;
+
+/*!\fn template <typename ... other_configs_t> configurator_base::configurator_base(configurator<other_configs_t...> const & cfg)
+ * \brief Constructs a new configurator_base object while delegating to the constructors of all inherited
+ *        sub-configurations.
+ * \memberof seqan3::detail::configurator_base
+ * \tparam remaining_configs_t A template parameter pack containing all the sub-configurations of the previous instance
+ *         of the configurator.
+ * \param cfg The previous configurator before extending it with a new configuration type.
+ *
+ * \details
+ *
+ * The constructor checks if there exists a common reference between `cfg` and the `remaining_configs_t` types.
+ * That is this class can only be constructed from `cfg` if and only if for every type in `remaining_configs_t`
+ * there is a corresponding type contained in `cfg`.
+ */
+//!\cond
+template <detail::config_concept config_head_t, detail::config_concept ... remaining_configs_t>
+class configurator_base<config_head_t, remaining_configs_t...> : public config_head_t,
+                                                                 public remaining_configs_t...
+{
+protected:
+
+    configurator_base()                                      = default;
+    configurator_base(configurator_base const &)             = default;
+    configurator_base(configurator_base &&)                  = default;
+    configurator_base & operator=(configurator_base const &) = default;
+    configurator_base & operator=(configurator_base &&)      = default;
+    ~configurator_base()                                     = default;
+
+    template <typename ... other_configs_t>
+        requires (common_reference_concept<configurator<other_configs_t ...>, remaining_configs_t> && ...)
+    constexpr explicit configurator_base(configurator<other_configs_t ...> const & cfg) :
+        remaining_configs_t{cfg}...
+    {}
+};
+//!\endcond
```

Comment:
Did you consider just implementing this as a std::tuple (or inheriting from std::tuple) and using the various std utilities like tuple_cat to tie and untie the config objects?

Reply:
At first I wanted to inherit the member functions of the config elements so that I can use the assembled configurator and call on this instance the member functions to set certain values. 
Consider that after initialisation some parameters might be changed. Like running the same config, but adapting the values of the band. But then I switched  to the `get` interface and afterwards I also thought to rather change it to the tuple implementation.